spark/src/test/scala/io/delta/sharing/spark/util/SchemaUtilsSuite.scala

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.sharing.spark.util
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.types._
+
+import io.delta.sharing.spark.util.SchemaUtils.isReadCompatible
+
+class SchemaUtilsSuite extends SparkFunSuite {
+  /////////////////////////////
+  // Read Compatibility Checks
+  /////////////////////////////
+
+  /**
+   * Tests change of datatype within a schema.
+   *  - the make() function is a "factory" function to create schemas that vary only by the
+   * given datatype in a specific position in the schema.
+   *  - other tests will call this method with different make() functions to test datatype
+   * incompatibility in all the different places within a schema (in a top-level struct,
+   * in a nested struct, as the element type of an array, etc.)
+   */
+  def testDatatypeChange(scenario: String)(make: DataType => StructType): Unit = {
+    val schemas = Map(
+      ("int", make(IntegerType)),
+      ("string", make(StringType)),
+      ("struct", make(new StructType().add("a", StringType))),
+      ("array", make(ArrayType(IntegerType))),
+      ("map", make(MapType(StringType, FloatType)))
+    )
+    test(s"change of datatype should fail read compatibility - $scenario") {
+      for (a <- schemas.keys; b <- schemas.keys if a != b) {
+        assert(!isReadCompatible(schemas(a), schemas(b)),
+          s"isReadCompatible should have failed for: ${schemas(a)}, ${schemas(b)}")
```

Comment:
Should we also test that if keys are equal then they are compatible?

Reply:
added